src/immer.d.ts

```diff
@@ -11,22 +14,38 @@
  * @returns The next state: a new state, or the current state if nothing was modified
  */
 export default function<S = any>(
-    currentState: S,
-    recipe: (this: S, draftState: S) => void
-): S
+    currentState: MaybeReadonly<S>,
+    recipe: (this: Readonly<S>, draftState: Draft<S>) => void
+): Readonly<S>
+
 // curried invocations
-export default function<S = any, A = any, B = any, C = any>(
-    recipe: (this: S, draftState: S, a: A, b: B, c: C) => void
-): (currentState: S, a: A, b: B, c: C) => S
-export default function<S = any, A = any, B = any>(
-    recipe: (this: S, draftState: S, a: A, b: B) => void
-): (currentState: S, a: A, b: B) => S
+
+// 0 additional arguments
+export default function<S = any>(
+    recipe: (this: Readonly<S>, draftState: Draft<S>) => void
```

Comment:
Shouldn't `this` be `Draft<S>` as well? (it's the same thing)

Reply:
Ah no because that would allow modification of the original state. With it being `Readonly<S>` the compiler complains if `this` is modified in the recipe.

```typescript
interface State { foo: string }
const first = { foo: "bar" };
const second = produce(state, draft => {
  this.foo = "can't do this, compiler error because 'this.foo' is readonly";
  draft.foo = "this works fine though";
});
second.foo = "this also produces a compiler error";
const third = produce(second, draft => {
  draft.foo = "but this does work, woo!";
});
```

The only time I can see this patch posing a problem is if someone defines their interfaces as completely readonly from the start. TS 2.8 is adding the ability to add/remove the readonly modifier when creating a new interface that references the type of a property of another interface (as in Draft/Pick/Readonly etc) but even then it wouldn't be possible to obtain the original readonly status.

As you said in your comment above the sensible thing to do is to define the interfaces in two steps. That also allows some properties to be readonly permanently such as:

```typescript
interface State {
  readonly appName: string;
  someValue: number;
}

const first: State = { appName: "My App", someValue: 0 };
const second = produce(first, draft => {
  draft.appName = "compiler error here";
  draft.someValue++; // fine
});
```

If there's a chance there are people already using immer who have made everything readonly from the start this would be a breaking change.

Reply:
Sorry @mweststrate when you said "they're the same" - they are **literally** the same object! My understanding was that `this` is the original state. So you could do:

```typescript
const original = { foo: 123, diff: 0 };
const next = produce(original, draft => {
  draft.foo += someValue;
  draft.diff = this.foo - draft.foo;
});
```

But that's not the case. Though perhaps it should be?